File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/internal/lifecycle/NativeHandles.kt
package org.maplibre.nativeffi.internal.lifecycle

import kotlin.jvm.JvmInline

/**
 * A handle the C API issued. The C API spells every handle as one integer type, so each kind gets
 * its own value class here to stay distinct at compile time. [raw] carries no ownership and is safe
 * to copy, compare, and hash. Zero is the null handle.
 */
internal sealed interface NativeHandle {
  val raw: Long

  val isNull: Boolean
    get() = raw == 0L
}

@JvmInline internal value class NativeRuntime(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeMap(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeMapProjection(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeRenderSession(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeOfflineRegionSnapshot(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeOfflineRegionList(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeOwnedBuffer(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeStyleIdList(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeStyleStringList(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeQueriedFeatureList(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeWakeSource(override val raw: Long) : NativeHandle

@JvmInline internal value class NativeResourceRequest(override val raw: Long) : NativeHandle

File: bindings/dart/lib/src/internal/lifecycle/native_handles.dart
/// A handle the C API issued.
///
/// The C API spells every handle as one integer type, so each kind gets its own
/// extension type here to keep the kinds distinct at compile time. The value
/// names one object for the life of the process, carries no ownership, and is
/// safe to copy, compare, hash, and send between isolates. Zero is the null
/// handle.
extension type const NativeHandle(int raw) implements Object {
  /// Whether this is the null handle.
  bool get isNull => raw == 0;
}

/// Runtime handle id.
extension type const NativeRuntime(int raw) implements NativeHandle {}

/// Map handle id.
extension type const NativeMap(int raw) implements NativeHandle {}

/// Map projection handle id.
extension type const NativeMapProjection(int raw) implements NativeHandle {}

/// Render session handle id.
extension type const NativeRenderSession(int raw) implements NativeHandle {}

/// Offline region snapshot handle id.
extension type const NativeOfflineRegionSnapshot(int raw)
    implements NativeHandle {}

/// Offline region list handle id.
extension type const NativeOfflineRegionList(int raw) implements NativeHandle {}

/// Style id list handle id.
extension type const NativeStyleIdList(int raw) implements NativeHandle {}

/// Style string list handle id.
extension type const NativeStyleStringList(int raw) implements NativeHandle {}

/// Wake source handle id.
extension type const NativeWakeSource(int raw) implements NativeHandle {}

/// Resource request handle id.
extension type const NativeResourceRequest(int raw) implements NativeHandle {}
